lib/tty/pager.rb

```diff
@@ -1,156 +1,79 @@
 # frozen_string_literal: true
 
-require 'tty-screen'
-
-require_relative 'pager/basic'
-require_relative 'pager/null'
-require_relative 'pager/system'
-require_relative 'pager/version'
+require_relative "pager/basic"
+require_relative "pager/null"
+require_relative "pager/system"
+require_relative "pager/version"
 
 module TTY
-  class Pager
+  module Pager
     Error = Class.new(StandardError)
     PagerClosed = Class.new(Error)
 
-    # Select an appriopriate pager
-    #
-    # If the user disabled paging then a NullPager is returned,
-    # otherwise a check is performed to find native system
-    # command to perform pagination with SystemPager. Finally,
-    # if no system command is found, a BasicPager is used which
-    # is a pure Ruby implementation known to work on any platform.
-    #
-    # @api private
-    def self.select_pager(enabled, commands)
-      if !enabled
-        NullPager
-      elsif SystemPager.exec_available?(*commands)
-        SystemPager
-      else
-        BasicPager
+    module ClassMethods
+      # Create a pager
+      #
+      # @param [Hash] options
+      # @option options [Proc] :prompt
+      #   a proc object that accepts page number
+      # @option options [IO] :input
+      #   the object to send input to
+      # @option options [IO] :output
+      #   the object to send output to
+      # @option options [Boolean] :enabled
+      #   disable/enable text paging
+      #
+      # @api public
+      def new(**options)
+        pager_klass = select_pager(**options)
+        @pager = pager_klass.new(**options)
       end
-    end
-
-    # Create a pager
-    #
-    # @param [Hash] options
-    # @option options [Proc] :prompt
-    #   a proc object that accepts page number
-    # @option options [IO] :input
-    #   the object to send input to
-    # @option options [IO] :output
-    #   the object to send output to
-    # @option options [Boolean] :enabled
-    #   disable/enable text paging
-    #
-    # @api public
-    def initialize(**options)
-      @input   = options.fetch(:input)  { $stdin }
-      @output  = options.fetch(:output) { $stdout }
-      @enabled = options.fetch(:enabled) { true }
-      commands = Array(options[:command])
 
-      if self.class == TTY::Pager
-        @pager = self.class.select_pager(@enabled, commands).new(**options)
+      # Paginate content through null, basic or system pager.
+      #
+      # @example
+      #   TTY::Pager.page do |pager|
+      #     pager.write "some text"
+      #   end
+      #
+      # @api public
+      def page(**options, &block)
+        @pager = new(**options)
+
+        begin
+          block.call(@pager)
+        rescue PagerClosed
+          # do nothing
+        ensure
+          @pager.close
+        end
       end
```

Comment:
Could the implementation of this method be `select_pager(**options).page(**options, &block)`? The `Abstract` class the same method, I think, so you can be sure every subclass will.

Reply:
A minor extra comment, for both this and the above comment -- it would be really nice if keyword args could be used. The constructors would be much clearer, and you wouldn't be sending unnecessary options. Though I imagine you'd like to keep maximum compatibility with older rubies. Still, I figure I should at least suggest it.

Reply:
It's not ideal to slurp up all the keyword args and pass them around. I'll see what I can do.

Reply:
Changed to use named arguments where possible, still had to slurp up few to pass from child to parent. But that's ok, it's a compromise. Overall, the readability is much better for each pager constructor.